### ISSUE TITLE
[5.1] [5.2] Connection::select can silently fail when using bound parameters in some situations

### ISSUE THREAD

**Issue by dark-panda** [laravel/framework#11189]
Take for instance the following:

```php
DB::table('foos')->whereRaw('1 = 1', [ 1 ])->count()
DB::table('foos')->whereRaw('1 = 1', [ 1 ])->get()
```

This will return `0` and an empty set, even when the table contains valid rows. No exception is raised, even though `PDOStatement::execute` has actually returned `false` and internally an error has been raise.

Generally PDO raises exceptions on most errors, but in this case it seems, either by design or by a bug, PHP doesn't raise an exception but does set the PDO error code and error info while returning `false` from `PDOStatement::execute`.

From https://github.com/laravel/framework/blob/5.1/src/Illuminate/Database/Connection.php#L321:

```php
$statement = $this->getPdoForSelect($useReadPdo)->prepare($query);

$statement->execute($me->prepareBindings($bindings));
```

`$statement->execute` will return `false` here for the above queries and the `select` function will basically silently return an empty set, leading to a count of 0 or an empty set of rows returned, even when there are valid rows in the `foos` table.

The reason for this is that in PHP's [ext/pdo/pdo_stmt.c](https://github.com/php/php-src/blob/master/ext/pdo/pdo_stmt.c), calls to the `stmt->methods->param_hook` callback around lines 388 and 414 are checked for errors as it returns 0 on error or 1 on success, but any errors set during those calls aren't propagated as exceptions via `pdo_raise_impl_error`. (In most cases that is; the PDO driver code is inconsistent at best on this matter.) 

By modifying the above code in Connection.php to dump out some debugging information like so:

```php
$ret = $statement->execute($me->prepareBindings($bindings));
var_dump($ret);
var_dump($statement->errorCode());
var_dump($statement->errorInfo());
```

We see the following for a valid query compared with the invalid query above:

```
# With both the PDO PostgreSQL and MySQL drivers on success:
bool(true)
string(5) "00000"
array(3) {
  [0]=>
  string(5) "00000"
  [1]=>
  NULL
  [2]=>
  NULL
}
```

Versus a bad query:

```
# With the PDO PostgreSQL driver on failure:
bool(false)
string(5) "00000"
array(3) {
  [0]=>
  string(5) "00000"
  [1]=>
  NULL
  [2]=>
  NULL
}

# With the PDO MySQL driver on failure:
bool(false)
string(5) "HY093"
array(3) {
  [0]=>
  string(5) "HY093"
  [1]=>
  NULL
  [2]=>
  NULL
}
```

I have not tested the other PDO drivers.

Unfortunately the PostgreSQL driver doesn't consistently set the error code. In some cases, like in the case where it's rewriting parameters from `$1`, `$2`, etc. around line 255, the code "HY093" is set on errors and `pdo_raise_impl_error` is even called, but in this particular case no error is raised and the error code is just set to "00000". This is obviously inconsistent behaviour in any event.

It would probably be best to check the return value of `$statement->execute` and raise an exception on a return value of `false` accordingly, as otherwise invalid statements can simply silently fail with no indication given as to why the query is returning an empty set or 0 or whatever the case may be. Now, this is potentially an error on the user's fault for writing a bad query in the first place, but the fact that there's a silent failure here regardless can make it difficult to track down that bad query.

**Comment by GrahamCampbell** [comment]
We're open to PRs. ;)

**Comment by themsaid** [comment]
Hello @dark-panda,
Any chance you managed to get around this issue or stand on the reasons it exists?

**Comment by reganjohnson** [comment]
@themsaid Perhaps related to: https://bugs.php.net/bug.php?id=70313

**Comment by derekmd** [comment]
As mentioned in @reganjohnson's comment, this was caused by the PDO PgSQL extension not raising the exception which has been fixed in PHP 5.6.26.

Now `QueryException` will be thrown when there is a `?` placeholder & bindings mismatch.

## Laravel [5.1][5.2] `whereIn()` throws `QueryException` on `null`

As a result, I now discovered a query builder `whereIn()` bug with:

* PHP 5.6.26
* pdo_pgsql
* Laravel 5.1 or 5.2

```php
DB::table('users')->whereIn('email', null)->get()
```

will throw:

```
Illuminate\Database\QueryException with message 'SQLSTATE[08P01]: <<Unknown error>>: 7 ERROR:  bind message supplies 1 parameters, but prepared statement "pdo_stmt_00000009" requires 0 (SQL: select * from "users" where 0 = 1)'
```

[Grammar@whereIn()](https://github.com/laravel/framework/blob/5.2/src/Illuminate/Database/Query/Grammars/Grammar.php#L372) applies:

```php
if (empty($where['values'])) {
    return '0 = 1';
}
```

And `Builder@whereIn()` ([5.1](https://github.com/laravel/framework/blob/5.1/src/Illuminate/Database/Query/Builder.php#L786) or [5.2](https://github.com/laravel/framework/blob/5.2/src/Illuminate/Database/Query/Builder.php#L916)) applies an extraneous `null` binding.

[5.3](https://github.com/laravel/framework/blob/5.3/src/Illuminate/Database/Query/Builder.php#L915) adding raw expressions causes:

```
PHP warning:  Invalid argument supplied for foreach() on line 915
```

In order to attain a desired noop expression, it looks bindings should only be applied when `!empty($values)` so the `Grammar@whereIn()` clause `0 = 1` doesn't cause a placeholder binding mismatch?

Likely a bug considered as a separate issue that affects every PDO extension.

**Comment by themsaid** [comment]
@derekmd I don't think `null` is a valid value in `whereIn` don't think it ever was in any release, you need to pass an Arrayable, Closure, or instance of builder.
